File: nine-language-version/lessons/08-higher-order-functions/hof.hs
-- Lesson 8: Higher-Order Functions in Haskell
--
-- Haskell is THE functional language - HOFs are everywhere!
-- - All functions are curried by default
-- - Function composition operator (.)
-- - Map, filter, fold built-in
-- - Lazy evaluation
-- - Type system helps catch errors

import Data.List (sort, sortBy)
import Data.Char (toUpper, toLower)

-- ====================
-- 1. Functions as First-Class Values
-- ====================

-- Functions are first-class - can be assigned, passed, returned
greet :: String -> String
greet name = "Hello, " ++ name ++ "!"

-- Store functions in a list
operations :: [Int -> Int]
operations = [(*2), (+1), (^2), (\x -> x - 1)]

applyAll :: [Int -> Int] -> Int -> [Int]
applyAll funcs x = map (\f -> f x) funcs

-- ====================
-- 2. Functions Taking Functions
-- ====================

applyTwice :: (a -> a) -> a -> a
applyTwice f x = f (f x)

applyNTimes :: Int -> (a -> a) -> a -> a
applyNTimes 0 _ x = x
applyNTimes n f x = applyNTimes (n - 1) f (f x)

-- ====================
-- 3. Functions Returning Functions (Currying)
-- ====================

-- All Haskell functions are curried by default!
add :: Int -> Int -> Int
add x y = x + y

-- Partial application is automatic
addFive :: Int -> Int
addFive = add 5  -- Partially applied!

-- Explicit function returning function
makeMultiplier :: Int -> (Int -> Int)
makeMultiplier n = \x -> x * n

makeAdder :: Int -> (Int -> Int)
makeAdder n = (+n)  -- Can use operator sections

-- ====================
-- 4. Map - Transform Each Element
-- ====================

-- map is built-in!
demonstrateMap :: IO ()
demonstrateMap = do
    let numbers = [1, 2, 3, 4, 5]

    -- Map with lambda
    putStrLn $ "   Doubled: " ++ show (map (*2) numbers)

    -- Map with named function
    putStrLn $ "   Squared: " ++ show (map (^2) numbers)

    -- Map multiple times (composition!)
    let result = map (*2) (map (+1) numbers)
    putStrLn $ "   Add 1 then double: " ++ show result

    -- Better: use composition
    let result2 = map ((*2) . (+1)) numbers
    putStrLn $ "   Same with composition: " ++ show result2

-- ====================
-- 5. Filter - Select Elements
-- ====================

-- filter is built-in!
demonstrateFilter :: IO ()
demonstrateFilter = do
    let numbers = [1..10]

    putStrLn $ "   Evens: " ++ show (filter even numbers)
    putStrLn $ "   Odds: " ++ show (filter odd numbers)
    putStrLn $ "   Greater than 5: " ++ show (filter (>5) numbers)

    -- Complex predicate
    let bigEvens = filter (\x -> even x && x > 5) numbers
    putStrLn $ "   Big evens: " ++ show bigEvens

-- ====================
-- 6. Fold (Reduce)
-- ====================

-- foldl = fold left, foldr = fold right
demonstrateFold :: IO ()
demonstrateFold = do
    let numbers = [1, 2, 3, 4, 5]

    -- Sum
    putStrLn $ "   Sum: " ++ show (foldl (+) 0 numbers)

    -- Product
    putStrLn $ "   Product: " ++ show (foldl (*) 1 numbers)

    -- Maximum
    putStrLn $ "   Max: " ++ show (foldl1 max numbers)

    -- Build a list backwards (demonstrating fold)
    let reversed = foldl (flip (:)) [] numbers
    putStrLn $ "   Reversed: " ++ show reversed

    -- Fold right vs left
    let words = ["Hello", "world", "from", "Haskell"]
    putStrLn $ "   foldr: " ++ foldr (\w acc -> w ++ " " ++ acc) "" words
    putStrLn $ "   foldl: " ++ foldl (\acc w -> acc ++ " " ++ w) "" words

-- ====================
-- 7. Closures (Captured Variables)
-- ====================

-- Haskell is pure, so "mutable closures" don't exist the same way
-- But we can return functions that capture values

makeCounter :: Int -> (Int -> Int)
makeCounter start = \x -> start + x

-- Using currying for closure-like behavior
addWithBase :: Int -> Int -> Int
addWithBase base value = base + value

-- ====================
-- 8. Currying and Partial Application
-- ====================

-- All functions are curried by default!
demonstrateCurrying :: IO ()
demonstrateCurrying = do
    -- Multi-argument function
    let power :: Int -> Int -> Int
        power base exp = base ^ exp

    -- Partial application
    let square = power 2
        cube = power 3

    putStrLn $ "   square 5 = " ++ show (square 5)
    putStrLn $ "   cube 5 = " ++ show (cube 5)

    -- Operator sections
    let addTen = (+10)
        divideByTwo = (/2)
        doubleIt = (*2)

    putStrLn $ "   addTen 5 = " ++ show (addTen 5)
    putStrLn $ "   doubleIt 5 = " ++ show (doubleIt 5)

-- ====================
-- 9. Function Composition
-- ====================

-- (.) is the composition operator!
demonstrateComposition :: IO ()
demonstrateComposition = do
    let addOne = (+1)
        double = (*2)
        square = (^2)

    -- Right-to-left composition
    let f = double . addOne
    putStrLn $ "   (double . addOne) 5 = " ++ show (f 5)

    -- Chain multiple
    let g = square . double . addOne
    putStrLn $ "   (square . double . addOne) 5 = " ++ show (g 5)

    -- Point-free style (no explicit argument)
    let processNumber = (\x -> x - 10) . square . double . addOne
    putStrLn $ "   Complex pipeline on 5 = " ++ show (processNumber 5)

    -- Composition with map
    let result = map (square . double) [1, 2, 3, 4, 5]
    putStrLn $ "   map (square . double) [1..5] = " ++ show result

-- ====================
-- 10. Common Higher-Order Functions
-- ====================

demonstrateCommon :: IO ()
demonstrateCommon = do
    let numbers = [1, 2, 3, 4, 5]

    -- all - all elements satisfy predicate
    putStrLn $ "   all (>0) numbers: " ++ show (all (>0) numbers)

    -- any - any element satisfies predicate
    putStrLn $ "   any even numbers: " ++ show (any even numbers)

    -- take - first n elements
    putStrLn $ "   take 3: " ++ show (take 3 numbers)

    -- drop - skip first n elements
    putStrLn $ "   drop 3: " ++ show (drop 3 numbers)

    -- takeWhile - take while predicate true
    putStrLn $ "   takeWhile (<4): " ++ show (takeWhile (<4) numbers)

    -- dropWhile - drop while predicate true
    putStrLn $ "   dropWhile (<4): " ++ show (dropWhile (<4) numbers)

    -- zip - combine two lists
    let list1 = [1, 2, 3]
        list2 = ['a', 'b', 'c']
    putStrLn $ "   zip: " ++ show (zip list1 list2)

    -- zipWith - combine with function
    putStrLn $ "   zipWith (+): " ++ show (zipWith (+) [1,2,3] [4,5,6])

-- ====================
-- 11. List Comprehensions (Alternative to HOFs)
-- ====================

demonstrateComprehensions :: IO ()
demonstrateComprehensions = do
    -- List comprehension = syntactic sugar for map/filter
    let numbers = [1..10]

    -- Equivalent to map (*2) numbers
    putStrLn $ "   [x*2 | x <- numbers]: " ++ show [x*2 | x <- numbers]

    -- Equivalent to filter even numbers
    putStrLn $ "   [x | x <- numbers, even x]: " ++ show [x | x <- numbers, even x]

    -- Combined map and filter
    putStrLn $ "   [x^2 | x <- numbers, x > 5]: " ++ show [x^2 | x <- numbers, x > 5]

-- ====================
-- 12. Custom HOFs
-- ====================

customMap :: (a -> b) -> [a] -> [b]
customMap _ [] = []
customMap f (x:xs) = f x : customMap f xs

customFilter :: (a -> Bool) -> [a] -> [a]
customFilter _ [] = []
customFilter p (x:xs)
    | p x       = x : customFilter p xs
    | otherwise = customFilter p xs

customFoldl :: (b -> a -> b) -> b -> [a] -> b
customFoldl _ acc [] = acc
customFoldl f acc (x:xs) = customFoldl f (f acc x) xs

-- ====================
-- 13. Higher-Order Functions with Type Classes
-- ====================

-- We can write very generic HOFs
twice :: (a -> a) -> a -> a
twice f = f . f

-- Works with any type!
demonstrateTwice :: IO ()
demonstrateTwice = do
    putStrLn $ "   twice (+1) 5 = " ++ show (twice (+1) 5)
    putStrLn $ "   twice (*2) 3 = " ++ show (twice (*2) 3)
    putStrLn $ "   twice reverse \"abc\" = " ++ twice reverse "abc"

-- ====================
-- 14. Flip - Swap Arguments
-- ====================

-- flip is built-in: reverses argument order
demonstrateFlip :: IO ()
demonstrateFlip = do
    let divide = (/)
        divideBy2 = flip divide 2  -- Now 2 is first argument

    putStrLn $ "   10 / 2 = " ++ show (divide 10 2)
    putStrLn $ "   divideBy2 10 = " ++ show (divideBy2 10)

    -- Useful with folds
    let numbers = [1, 2, 3, 4, 5]
        reversed = foldl (flip (:)) [] numbers
    putStrLn $ "   foldl (flip (:)) [] = " ++ show reversed

-- ====================
-- 15. Real-World Example: Data Pipeline
-- ====================

data Person = Person { name :: String, age :: Int } deriving (Show)

processData :: [Person] -> [Person]
processData = take 10                           -- Top 10
            . sortBy (\a b -> compare (age b) (age a))  -- Sort by age desc
            . map (\p -> p { name = toLowerName (name p) })  -- Normalize
            . filter ((>= 18) . age)            -- Adults only
    where
        toLowerName = map toLower

-- Point-free style!
sumOfSquaresOfPositives :: [Int] -> Int
sumOfSquaresOfPositives = sum . map (^2) . filter (>0)

-- ====================
-- Main Demonstration
-- ====================

main :: IO ()
main = do
    putStrLn "=== Higher-Order Functions in Haskell ===\n"

    -- 1. First-class functions
    putStrLn "1. Functions as First-Class Values:"
    putStrLn $ "   applyAll operations 5 = " ++ show (applyAll operations 5)

    -- 2. Functions taking functions
    putStrLn "\n2. Functions Taking Functions:"
    putStrLn $ "   applyTwice (+1) 5 = " ++ show (applyTwice (+1) 5)
    putStrLn $ "   applyNTimes 3 (*2) 2 = " ++ show (applyNTimes 3 (*2) 2)

    -- 3. Currying
    putStrLn "\n3. Currying (Automatic in Haskell!):"
    let timesThree = makeMultiplier 3
    putStrLn $ "   makeMultiplier 3 7 = " ++ show (timesThree 7)
    putStrLn $ "   addFive 10 = " ++ show (addFive 10)

    -- 4. Map
    putStrLn "\n4. Map - Transform Each Element:"
    demonstrateMap

    -- 5. Filter
    putStrLn "\n5. Filter - Select Elements:"
    demonstrateFilter

    -- 6. Fold
    putStrLn "\n6. Fold - Combine to Single Value:"
    demonstrateFold

    -- 7. Currying and partial application
    putStrLn "\n7. Currying and Partial Application:"
    demonstrateCurrying

    -- 8. Function composition
    putStrLn "\n8. Function Composition:"
    demonstrateComposition

    -- 9. Common HOFs
    putStrLn "\n9. Common Higher-Order Functions:"
    demonstrateCommon

    -- 10. List comprehensions
    putStrLn "\n10. List Comprehensions:"
    demonstrateComprehensions

    -- 11. Custom implementations
    putStrLn "\n11. Custom HOF Implementations:"
    let numbers = [1, 2, 3, 4, 5]
    putStrLn $ "   customMap (*2): " ++ show (customMap (*2) numbers)
    putStrLn $ "   customFilter even: " ++ show (customFilter even numbers)
    putStrLn $ "   customFoldl (+) 0: " ++ show (customFoldl (+) 0 numbers)

    -- 12. Twice
    putStrLn "\n12. Generic twice Function:"
    demonstrateTwice

    -- 13. Flip
    putStrLn "\n13. Flip - Reverse Arguments:"
    demonstrateFlip

    -- 14. Real-world example
    putStrLn "\n14. Real-World Data Pipeline:"
    let people = [Person "alice" 25, Person "BOB" 17, Person "charlie" 30]
        processed = processData people
    putStrLn $ "   Processed: " ++ show processed

    putStrLn "\n15. Point-Free Style:"
    putStrLn $ "   sumOfSquaresOfPositives [-1,2,3,-4,5] = " ++
               show (sumOfSquaresOfPositives [-1,2,3,-4,5])

    putStrLn "\n16. Key Haskell Features:"
    putStrLn "   - All functions are curried by default"
    putStrLn "   - Composition operator (.) for elegant pipelines"
    putStrLn "   - Point-free style (no explicit arguments)"
    putStrLn "   - Lazy evaluation (infinite lists possible!)"
    putStrLn "   - Strong type system catches errors"

File: nine-language-version/lessons/08-higher-order-functions/hof.pl
% Lesson 8: Higher-Order Functions in Prolog
%
% Prolog is logic-based, not functional, but still supports HOFs!
% - Predicates can take other predicates as arguments
% - maplist, include, exclude are built-in HOFs
% - call/N for calling predicates dynamically
% - findall, bagof, setof for collecting solutions
%
% Different paradigm, similar concepts!

:- initialization(main).

% ====================
% 1. Predicates as Arguments
% ====================

% Apply a unary predicate twice
apply_twice(Pred, X, Result) :-
    call(Pred, X, Temp),
    call(Pred, Temp, Result).

% Apply n times
apply_n_times(_, 0, X, X).
apply_n_times(Pred, N, X, Result) :-
    N > 0,
    call(Pred, X, Temp),
    N1 is N - 1,
    apply_n_times(Pred, N1, Temp, Result).

% Example predicates
add_one(X, Y) :- Y is X + 1.
double_it(X, Y) :- Y is X * 2.
square_it(X, Y) :- Y is X * X.

% ====================
% 2. Maplist - Transform Each Element
% ====================

demonstrate_maplist :-
    Numbers = [1, 2, 3, 4, 5],

    % maplist/3 - apply predicate to each element
    maplist(double_it, Numbers, Doubled),
    format('   Doubled: ~w~n', [Doubled]),

    % maplist/3 with lambda-like syntax (library)
    maplist([X,Y]>>(Y is X * X), Numbers, Squared),
    format('   Squared: ~w~n', [Squared]).

% ====================
% 3. Include/Exclude - Filter Elements
% ====================

is_even(X) :- 0 is X mod 2.
is_odd(X) :- 1 is X mod 2.
is_positive(X) :- X > 0.

demonstrate_filter :-
    Numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],

    % include = filter (keep matching)
    include(is_even, Numbers, Evens),
    format('   Evens: ~w~n', [Evens]),

    % exclude = inverse filter
    exclude(is_even, Numbers, Odds),
    format('   Odds: ~w~n', [Odds]),

    % Complex predicate
    include([X]>>(is_odd(X), X > 5), Numbers, BigOdds),
    format('   Big odds: ~w~n', [BigOdds]).

% ====================
% 4. Foldl - Reduce/Fold
% ====================

% Sum accumulator
sum_acc(X, Acc, Result) :- Result is Acc + X.

% Product accumulator
product_acc(X, Acc, Result) :- Result is Acc * X.

% Max accumulator
max_acc(X, Acc, Result) :- Result is max(X, Acc).

demonstrate_fold :-
    Numbers = [1, 2, 3, 4, 5],

    % foldl(Pred, List, Initial, Final)
    foldl(sum_acc, Numbers, 0, Sum),
    format('   Sum: ~w~n', [Sum]),

    foldl(product_acc, Numbers, 1, Product),
    format('   Product: ~w~n', [Product]),

    foldl(max_acc, Numbers, 0, Max),
    format('   Max: ~w~n', [Max]).

% ====================
% 5. Higher-Order Predicates
% ====================

% Apply predicate to all elements and collect results
apply_to_all(_, [], []).
apply_to_all(Pred, [H|T], [RH|RT]) :-
    call(Pred, H, RH),
    apply_to_all(Pred, T, RT).

% Filter with predicate
filter_list(_, [], []).
filter_list(Pred, [H|T], [H|RT]) :-
    call(Pred, H),
    !,
    filter_list(Pred, T, RT).
filter_list(Pred, [_|T], RT) :-
    filter_list(Pred, T, RT).

% ====================
% 6. Findall - Collecting Solutions
% ====================

% Define some facts
person(alice, 30).
person(bob, 25).
person(charlie, 35).
person(diana, 28).

age_category(Age, Category) :-
    (Age >= 30 -> Category = adult ; Category = young).

demonstrate_findall :-
    % Collect all names
    findall(Name, person(Name, _), Names),
    format('   All names: ~w~n', [Names]),

    % Collect ages over 25
    findall(Age, (person(_, Age), Age > 25), Ages),
    format('   Ages > 25: ~w~n', [Ages]),

    % Complex query with transformation
    findall(Category, (person(_, Age), age_category(Age, Category)), Categories),
    format('   Categories: ~w~n', [Categories]).

% ====================
% 7. Composition
% ====================

% Compose two predicates
compose(F, G, X, Z) :-
    call(G, X, Y),
    call(F, Y, Z).

demonstrate_composition :-
    compose(double_it, add_one, 5, Result),
    format('   compose(double, add_one, 5) = ~w~n', [Result]).

% ====================
% 8. Currying Simulation
% ====================

% Partial application simulation
multiply(X, Y, Result) :- Result is X * Y.

% Create a "specialized" version
times_three(Y, Result) :- multiply(3, Y, Result).
times_five(Y, Result) :- multiply(5, Y, Result).

demonstrate_currying :-
    times_three(7, R1),
    times_five(4, R2),
    format('   times_three(7) = ~w~n', [R1]),
    format('   times_five(4) = ~w~n', [R2]).

% ====================
% 9. All/Any (Like all/any in other languages)
% ====================

all_satisfy(_, []).
all_satisfy(Pred, [H|T]) :-
    call(Pred, H),
    all_satisfy(Pred, T).

any_satisfy(Pred, [H|_]) :-
    call(Pred, H), !.
any_satisfy(Pred, [_|T]) :-
    any_satisfy(Pred, T).

demonstrate_all_any :-
    Numbers = [1, 2, 3, 4, 5],

    (all_satisfy(is_positive, Numbers) ->
        format('   All positive: true~n') ;
        format('   All positive: false~n')),

    (any_satisfy(is_even, Numbers) ->
        format('   Any even: true~n') ;
        format('   Any even: false~n')).

% ====================
% 10. Partition - Split List
% ====================

partition_list(_, [], [], []).
partition_list(Pred, [H|T], [H|True], False) :-
    call(Pred, H),
    !,
    partition_list(Pred, T, True, False).
partition_list(Pred, [H|T], True, [H|False]) :-
    partition_list(Pred, T, True, False).

demonstrate_partition :-
    Numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
    partition_list(is_even, Numbers, Evens, Odds),
    format('   Partition evens: ~w~n', [Evens]),
    format('   Partition odds: ~w~n', [Odds]).

% ====================
% 11. Custom Map Implementation
% ====================

custom_map(_, [], []).
custom_map(Pred, [H|T], [RH|RT]) :-
    call(Pred, H, RH),
    custom_map(Pred, T, RT).

custom_filter(_, [], []).
custom_filter(Pred, [H|T], [H|RT]) :-
    call(Pred, H),
    !,
    custom_filter(Pred, T, RT).
custom_filter(Pred, [_|T], RT) :-
    custom_filter(Pred, T, RT).

custom_foldl(_, [], Acc, Acc).
custom_foldl(Pred, [H|T], Acc, Result) :-
    call(Pred, H, Acc, NewAcc),
    custom_foldl(Pred, T, NewAcc, Result).

demonstrate_custom :-
    Numbers = [1, 2, 3, 4, 5],

    custom_map(double_it, Numbers, Doubled),
    format('   custom_map: ~w~n', [Doubled]),

    custom_filter(is_even, Numbers, Evens),
    format('   custom_filter: ~w~n', [Evens]),

    custom_foldl(sum_acc, Numbers, 0, Sum),
    format('   custom_foldl: ~w~n', [Sum]).

% ====================
% 12. Real-World Example: Data Processing
% ====================

% Process user data
process_user([Name, Age], [ProcessedName, Category]) :-
    downcase_atom(Name, ProcessedName),
    (Age >= 18 -> Category = adult ; Category = minor).

demonstrate_pipeline :-
    Users = [[' ALICE ', 25], ['BOB', 17], ['CHARLIE ', 30]],

    % Filter active users (all in this case)
    % Map to process
    maplist(process_user, Users, Processed),
    format('   Processed users: ~w~n', [Processed]).

% ====================
% Main Demonstration
% ====================

main :-
    writeln('=== Higher-Order Functions in Prolog ===\n'),

    % 1. Predicates as arguments
    writeln('1. Predicates as Arguments:'),
    apply_twice(add_one, 5, R1),
    format('   apply_twice(add_one, 5) = ~w~n', [R1]),
    apply_n_times(double_it, 3, 2, R2),
    format('   apply_n_times(double_it, 3, 2) = ~w~n', [R2]),

    % 2. Maplist
    nl, writeln('2. Maplist - Transform Each Element:'),
    demonstrate_maplist,

    % 3. Include/Exclude
    nl, writeln('3. Include/Exclude - Filter Elements:'),
    demonstrate_filter,

    % 4. Foldl
    nl, writeln('4. Foldl - Combine to Single Value:'),
    demonstrate_fold,

    % 5. Findall
    nl, writeln('5. Findall - Collect Solutions:'),
    demonstrate_findall,

    % 6. Composition
    nl, writeln('6. Function Composition:'),
    demonstrate_composition,

    % 7. Currying
    nl, writeln('7. Partial Application (Currying):'),
    demonstrate_currying,

    % 8. All/Any
    nl, writeln('8. All/Any Predicates:'),
    demonstrate_all_any,

    % 9. Partition
    nl, writeln('9. Partition:'),
    demonstrate_partition,

    % 10. Custom implementations
    nl, writeln('10. Custom HOF Implementations:'),
    demonstrate_custom,

    % 11. Pipeline
    nl, writeln('11. Real-World Data Pipeline:'),
    demonstrate_pipeline,

    nl, writeln('12. Prolog\'s Unique Approach:'),
    writeln('   - Predicates as arguments (not functions)'),
    writeln('   - call/N for dynamic predicate calling'),
    writeln('   - maplist, include, exclude built-in'),
    writeln('   - findall for collecting solutions'),
    writeln('   - Logic-based, but supports functional concepts!'),

    halt(0).

:- catch(main, E, (writeln(E), halt(1))).
